src/Parser.php

```diff
@@ -363,18 +358,37 @@ protected function isIpv6Host(string $ipv6): bool
      */
     protected function isRegisteredName(string $host): bool
     {
-        if ('.' === substr($host, -1, 1)) {
-            $host = substr($host, 0, -1);
+        // Note that unreserved is purposely missing . and -
+        // . is used to separate labels
+        // - cannot start or end a label
+        static $pattern = '/(?(DEFINE)
+                (?<unreserved> [a-z0-9\_\~])
+                (?<gen_delims> [:\/?#\[\]@])
+                (?<sub_delims> [!$&\'()*+,;=])
+                (?<label> (?:(?&unreserved)(?:(?&unreserved)|(?&gen_delims)|(?&sub_delims)|\-){0,61})?(?&unreserved))
+            )
+            ^(?:(?&label)\.){0,126}(?&label)\.?$/imx';
+
+        if (preg_match($pattern, $host)) {
+            return true;
         }
 
-        $labels = array_map([$this, 'toAscii'], explode('.', $host));
+        if (false !== strpos($host, '%')) {
+            $host = rawurldecode($host);
+        }
 
-        return 127 > count($labels)
-            && 253 > strlen(implode('.', $labels))
-            && $labels === array_filter($labels, [$this, 'isHostLabel']);
+        $host = idn_to_ascii($host, IDNA_NONTRANSITIONAL_TO_ASCII, INTL_IDNA_VARIANT_UTS46);
+
+        if (false === $host) {
+            return false;
+        }
+
+        return (bool) preg_match($pattern, $host);
     }
 
     /**
+     * @deprecated Will be removed in v2.0
```

Comment:
@nyamsprod Should we trigger deprecation notices using `trigger_error` or not?

Reply:
@kelunik this method is internal so unless someone does extends this class (which I highly doubt) I don't see the advantage of using `trigger_error`. Marking it deprecated is enough for me

Reply:
@nyamsprod If it's intended that way, that's a perfect reason to use `trigger_error`. As we don't use it anymore, any users can be alerted that it will be made private and the complete class be made final in a future release.

Reply:
fine by me

Reply:
Added `trigger_error` to these methods.